File: built-in plugins dsp/mb-granular-synth.cpp
/**
 * MB Granular Synth
 * Category : instrument
 * Type     : wavetable
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Advanced granular synthesizer for textural and ambient sounds
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_GRANULAR_SYNTH_H
#define MB_GRANULAR_SYNTH_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbGranularSynth : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-granular-synth";
    static constexpr const char* PLUGIN_NAME    = "MB Granular Synth";
    static constexpr const char* PLUGIN_TYPE    = "wavetable";
    static constexpr const char* PLUGIN_CATEGORY = "instrument";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float grain_size = 50f;  // range [1, 500]
    float grain_density = 20f;  // range [1, 100]
    float position = 0f;  // range [0, 1]
    float position_spread = 0.1f;  // range [0, 1]
    float pitch_spread = 0f;  // range [-24, 24]
    float pan_spread = 0.5f;  // range [0, 1]
    float reverse = 0f;  // range [0, 1]
    float volume = 0.8f;  // range [0, 1]
    };

    MbGranularSynth() = default;
    ~MbGranularSynth() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.grain_size = std::clamp(params.grain_size, 1f, 500f);
        params.grain_density = std::clamp(params.grain_density, 1f, 100f);
        params.position = std::clamp(params.position, 0f, 1f);
        params.position_spread = std::clamp(params.position_spread, 0f, 1f);
        params.pitch_spread = std::clamp(params.pitch_spread, -24f, 24f);
        params.pan_spread = std::clamp(params.pan_spread, 0f, 1f);
        params.reverse = std::clamp(params.reverse, 0f, 1f);
        params.volume = std::clamp(params.volume, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Granular Synth
        return input;
    }
};

#endif // MB_GRANULAR_SYNTH_H

File: built-in plugins dsp/mb-granular-synth.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbGranularSynthPlugin: PluginDefinition = {
    id: 'mb-granular-synth',
    slug: 'mb-granular-synth',
    name: 'MB Granular Synth',
    category: 'instrument',
    type: 'wavetable',
    version: '1.0.0',
    description: 'Advanced granular synthesizer for textural and ambient sounds',
    author: 'Max Booster', grade: 'A',
    oscillators: [
      { type: 'sine', detune: 0, gain: 0.5 },
      { type: 'triangle', detune: 0, gain: 0.3 },
    ],
    envelope: { attack: 0.5, decay: 1.0, sustain: 0.7, release: 2.0 },
    parameters: [
      { id: 'grain_size', name: 'Grain Size', type: 'float', defaultValue: 50, minValue: 1, maxValue: 500, unit: 'ms', automatable: true },
      { id: 'grain_density', name: 'Grain Density', type: 'float', defaultValue: 20, minValue: 1, maxValue: 100, unit: 'grains/s', automatable: true },
      { id: 'position', name: 'Sample Position', type: 'float', defaultValue: 0, minValue: 0, maxValue: 1, automatable: true },
      { id: 'position_spread', name: 'Position Spread', type: 'float', defaultValue: 0.1, minValue: 0, maxValue: 1, automatable: true },
      { id: 'pitch_spread', name: 'Pitch Spread', type: 'float', defaultValue: 0, minValue: -24, maxValue: 24, unit: 'semitones', automatable: true },
      { id: 'pan_spread', name: 'Pan Spread', type: 'float', defaultValue: 0.5, minValue: 0, maxValue: 1, automatable: true },
      { id: 'reverse', name: 'Reverse Probability', type: 'float', defaultValue: 0, minValue: 0, maxValue: 1, automatable: true },
      { id: 'freeze', name: 'Freeze', type: 'bool', defaultValue: false, automatable: true },
      { id: 'volume', name: 'Volume', type: 'float', defaultValue: 0.8, minValue: 0, maxValue: 1, automatable: true },
    ],
    defaultPreset: { grain_size: 50, grain_density: 20, position: 0, position_spread: 0.1, pitch_spread: 0, pan_spread: 0.5, reverse: 0, freeze: false, volume: 0.8 },
  };

export default MbGranularSynthPlugin;
